numpy/core/src/multiarray/methods.c

```diff
@@ -1426,21 +1429,44 @@ array_deepcopy(PyArrayObject *self, PyObject *args)
         if (deepcopy == NULL) {
             return NULL;
         }
-        it = (PyArrayIterObject *)PyArray_IterNew((PyObject *)self);
```

Comment:
The problem is that IterNew is iterating in C order, and optr is therefore getting out of synch.